src/StoreApi/Utilities/ProductQueryFilters.php

```diff
@@ -111,63 +113,194 @@ private function generate_stock_status_count_query( $status, $product_query_sql,
-		$product_query_sql            = $query->request;
+	/**
+	 * Get attribute and meta counts.
+	 *
+	 * @param WP_REST_Request $request Request data.
+	 * @param string          $filtered_attribute The attribute to count.
+	 *
+	 * @return array
+	 */
+	public function get_attribute_counts( $request, $filtered_attribute ) {
+		if ( is_array( $filtered_attribute ) ) {
+			wc_deprecated_argument( 'attributes', 'TBD', 'get_attribute_counts does not require an array of attributes as the second parameter anymore. Provide the filtered attribute as a string instead.' );
 
-		remove_filter( 'posts_clauses', array( $product_query, 'add_query_clauses' ), 10 );
-		remove_filter( 'posts_pre_query', '__return_empty_array' );
+			$filtered_attribute = ! empty( $filtered_attribute[0] ) ? $filtered_attribute[0] : '';
 
-		if ( count( $attributes ) === count( array_filter( $attributes, 'is_numeric' ) ) ) {
-			$attributes = array_map( 'wc_attribute_taxonomy_name_by_id', wp_parse_id_list( $attributes ) );
+			if ( empty( $filtered_attribute ) ) {
+				return array();
+			}
 		}
 
-		$attributes_to_count     = array_map(
-			function( $attribute ) {
-				$attribute = wc_sanitize_taxonomy_name( $attribute );
-				return esc_sql( $attribute );
-			},
-			$attributes
+		$attributes_data            = $request->get_param( 'attributes' );
+		$calculate_attribute_counts = $request->get_param( 'calculate_attribute_counts' );
+		$min_price                  = $request->get_param( 'min_price' );
+		$max_price                  = $request->get_param( 'max_price' );
+		$rating                     = $request->get_param( 'rating' );
+		$stock_status               = $request->get_param( 'stock_status' );
+
+		$transient_key = 'wc_get_attribute_and_meta_counts_' . md5(
+			wp_json_encode(
+				array(
+					'attributes_data'            => $attributes_data,
+					'calculate_attribute_counts' => $calculate_attribute_counts,
+					'min_price'                  => $min_price,
+					'max_price'                  => $max_price,
+					'rating'                     => $rating,
+					'stock_status'               => $stock_status,
+					'filtered_attribute'         => $filtered_attribute,
+				)
+			)
 		);
-		$attributes_to_count_sql = 'AND term_taxonomy.taxonomy IN ("' . implode( '","', $attributes_to_count ) . '")';
-		$attribute_count_sql     = "
-			SELECT COUNT( DISTINCT posts.ID ) as term_count, terms.term_id as term_count_id
-			FROM {$wpdb->posts} AS posts
-			INNER JOIN {$wpdb->term_relationships} AS term_relationships ON posts.ID = term_relationships.object_id
-			INNER JOIN {$wpdb->term_taxonomy} AS term_taxonomy USING( term_taxonomy_id )
-			INNER JOIN {$wpdb->terms} AS terms USING( term_id )
-			WHERE posts.ID IN ( {$product_query_sql} )
-			{$attributes_to_count_sql}
-			GROUP BY terms.term_id
-		";
 
-		$results = $wpdb->get_results( $attribute_count_sql ); // phpcs:ignore
+		$cached_results = get_transient( $transient_key );
+
+		if ( ! empty( $cached_results ) ) {
```

Comment:
For the caching, what do you think about adding a check for `WP_DEBUG` constant? This way developers can test without having to comment out the `return` and prevent false negatives. It is common for developers to set `WP_DEBUG` to `true` in development environment so we could use this to our advantage especially when testing.

`if ( ! empty( $cached_results ) && ( defined( 'WP_DEBUG' ) && ! WP_DEBUG ) ) {`

Reply:
Good idea! [Done](https://github.com/woocommerce/woocommerce-blocks/pull/8599/commits/497e101a5f58a7bfc82cf3415b33ab05af925a31).